README.md
# ExpandTapArea

Manually expand the tap area of a widget without changing its size or layout. Similar with the *hitTest* in iOS development.

# Context

```
In the Human Interface Guidelines, Apple recommends a minimum target size of 44 x 44 px. 
In the Android Material Design Guidelines, it’s suggested that touch targets should be at least 48 x 48 dp.
```
Inspired by [issue](https://github.com/flutter/flutter/issues/31728)

# Usage

```dart
ExpandTapWidget(
  onTap: () {},
  tapPadding: EdgeInsets.all(25.0),
  child: smallWidget(),
),
```

# Demo

<div style="text-align: center">
    <table>
        <tr>
            <td style="text-align: center">
                <a href="https://raw.githubusercontent.com/malikwang/expand_tap_area/master/images/1.png">
                    <img src="https://raw.githubusercontent.com/malikwang/expand_tap_area/master/images/1.png" width="300"/>
                </a>
            </td>            
            <td style="text-align: center">
                <a href="https://raw.githubusercontent.com/malikwang/expand_tap_area/master/images/2.png">
                    <img src="https://raw.githubusercontent.com/malikwang/expand_tap_area/master/images/2.png" width="300"/>
                </a>
            </td>  
        </tr>
    </table>
</div>

# Attention

1. ExpandTapWidget can be wrapped inside/outside GestureDetector: **To avoid the conflict of gesture, it is better to apply on the innermost/smallest widget.**
2.  **The expanded area wouldn't exceed the boundary of the Parent Box. You can set *debugPaintExpandAreaEnabled = true* to display the actual expanded area.**



CHANGELOG.md
## 1.1.1

* Fix Flutter 3.0.0 warning.

## 1.1.0

* Null safety release.

## 1.0.1

* Add debugPaintExpandAreaEnabled flag.

## 1.0.0

* Initial open source release.


lib/src/expand_tap_area.dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'dart:ui';
import 'package:flutter/foundation.dart';

bool debugPaintExpandAreaEnabled = false;
Color debugPaintExpandAreaColor = const Color(0xFFFF0000).withOpacity(0.4);

// If there is someone who knows how to extend the hitarea above the parent rect, so it is not
// Clipped, please add that here. This would be really helpful.
Color debugPaintClipAreaColor = const Color(0xFF0000FF).withOpacity(0.4);

class ExpandTapWidget extends SingleChildRenderObjectWidget {
  ExpandTapWidget({
    Key? key,
    Widget? child,
    required this.onTap,
    required this.tapPadding,
  }) : super(key: key, child: child);

  final VoidCallback onTap;
  final EdgeInsets tapPadding;

  @override
  RenderObject createRenderObject(BuildContext context) => _ExpandTapRenderBox(
    onTap: onTap,
    tapPadding: tapPadding,
  );
}

class _TmpGestureArenaMember extends GestureArenaMember {
  _TmpGestureArenaMember({
    required this.onTap,
  });

  final VoidCallback onTap;

  @override
  void acceptGesture(int key) {
    this.onTap();
  }

  @override
  void rejectGesture(int key) {}
}

class _ExpandTapRenderBox extends RenderBox
    with RenderObjectWithChildMixin<RenderBox> {
  _ExpandTapRenderBox({
    required this.onTap,
    this.tapPadding = EdgeInsets.zero,
  });

  final VoidCallback onTap;
  final EdgeInsets tapPadding;

  @override
  void performLayout() {
    child!.layout(constraints, parentUsesSize: true);
    size = child!.size;
  }

  @override
  void paint(PaintingContext context, Offset offset) {
    if (child != null) {
      final BoxParentData childParentData = child!.parentData! as BoxParentData;
      context.paintChild(child!, childParentData.offset + offset);
    }
    if (debugPaintExpandAreaEnabled) {
      debugPaintExpandArea(context, offset);
    }
  }

  void debugPaintExpandArea(PaintingContext context, Offset offset) {
    RenderBox parentBox = parent as RenderBox;

    /// [parentPosition]: the offset of current renderbox's parent
    Offset parentPosition = Offset.zero;
    parentPosition = offset - localToGlobal(Offset.zero, ancestor: parentBox);

    /// we directly get this offset using the method [localToGlobal] of parentBox, like the following annotated code
    /// However, this result is inaccurate, you can uncomment to try out.
    // parentPosition = parentBox.localToGlobal(Offset.zero);

    Size parentSize = parentBox.size;
    Rect parentRect = Rect.fromLTWH(
      parentPosition.dx,
      parentPosition.dy,
      parentSize.width,
      parentSize.height,
    );
    final BoxParentData childParentData = child!.parentData! as BoxParentData;
    Offset paintOffset = childParentData.offset + offset - tapPadding.topLeft;
    Rect paintRect = Rect.fromLTWH(
      paintOffset.dx,
      paintOffset.dy,
      size.width + tapPadding.horizontal,
      size.height + tapPadding.vertical,
    );
      final Paint paint = Paint()
        ..style = PaintingStyle.fill
        ..strokeWidth = 1.0
        ..color = debugPaintExpandAreaColor;

      final Paint paint2 = Paint()
        ..style = PaintingStyle.fill
        ..strokeWidth = 1.0
        ..color = debugPaintClipAreaColor;
      context.canvas.drawRect(paintRect, paint);
      context.canvas.drawRect(paintRect.intersect(parentRect), paint2);
  }


  @override
  void handleEvent(PointerEvent event, BoxHitTestEntry entry) {
    if (event is PointerDownEvent) {
      _TmpGestureArenaMember member = _TmpGestureArenaMember(onTap: onTap);
      GestureBinding.instance.gestureArena.add(event.pointer, member);
    } else if (event is PointerUpEvent) {
      GestureBinding.instance.gestureArena.sweep(event.pointer);
    }
  }

  @override
  bool hitTestSelf(Offset position) => true;

  @override
  bool hitTestChildren(BoxHitTestResult result, {Offset? position}) {
    this.visitChildren((child) {
      if (child is RenderBox) {
        final BoxParentData parentData = child.parentData! as BoxParentData;
        if (child.hitTest(result, position: position! - parentData.offset)) {
         return;
        }
      }
    });
    return false;
  }

  @override
  bool hitTest(BoxHitTestResult result, {Offset? position}) {
    Rect expandRect = Rect.fromLTWH(
      0 - tapPadding.left,
      0 - tapPadding.top,
      size.width + tapPadding.right + tapPadding.left,
      size.height + tapPadding.top + tapPadding.bottom,
    );
    if (expandRect.contains(position!)) {
      bool hitTarget =
          hitTestChildren(result, position: position) || hitTestSelf(position);
      if (hitTarget) {
        result.add(BoxHitTestEntry(this, position));
        return true;
      }
    }
    return false;
  }
}


example/lib/main.dart
import 'package:expand_tap_area/expand_tap_area.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:fluttertoast/fluttertoast.dart';

void main() {
  debugPaintExpandAreaEnabled = true;
  debugPaintExpandAreaColor = Color(0xFF00FFFF).withOpacity(0.2);
  runApp(MyApp());
}

_showToast(String msg) {
  Fluttertoast.showToast(
    msg: msg,
    toastLength: Toast.LENGTH_SHORT,
    gravity: ToastGravity.CENTER,
    textColor: Colors.white,
    fontSize: 16.0,
  );
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'ExpandTapArea',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'ExpandTapArea'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  Widget _buildLarge() {
    return GestureDetector(
      onTap: () => _showToast('44 is Tapped'),
      child: Container(
        width: 44,
        height: 44,
        color: Colors.greenAccent,
        alignment: Alignment.center,
        child: Text(
          '44',
          style: TextStyle(
            color: Colors.white,
          ),
        ),
      ),
    );
  }

  Widget _buildMedium() {
    return GestureDetector(
      onTap: () => _showToast('30 is Tapped'),
      child: Container(
        width: 30,
        height: 30,
        color: Colors.orangeAccent,
        alignment: Alignment.center,
        child: Text(
          '30',
          style: TextStyle(
            color: Colors.white,
          ),
        ),
      ),
    );
  }

  Widget _buildSmall() {
    return Container(
      width: 22,
      height: 22,
      color: Colors.redAccent,
      alignment: Alignment.center,
      child: Text(
        '22',
        style: TextStyle(
          color: Colors.white,
        ),
      ),
    );
  }

  Widget _buildCell1() {
    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 10),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(
              'In the Human Interface Guidelines, Apple recommends a minimum target size of 44 x 44 px. '),
          Text(
              'In the Android Material Design Guidelines, it’s suggested that touch targets should be at least 48 x 48 dp.'),
          Padding(
            padding: const EdgeInsets.only(top: 10),
            child: Container(
              decoration: BoxDecoration(
                borderRadius: BorderRadius.all(
                  Radius.circular(5),
                ),
                border: Border.all(
                  color: Colors.black,
                  width: 1,
                ),
              ),
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: [
                  _buildLarge(),
                  _buildMedium(),
                  GestureDetector(
                    onTap: () => _showToast('22 is Tapped'),
                    child: _buildSmall(),
                  ),
                ],
              ),
            ),
          ),
          Padding(
            padding: const EdgeInsets.only(top: 10),
            child: Text('The black border is parent box size.'),
          ),
        ],
      ),
    );
  }

  Widget _buildCell2() {
    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 10),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(
              'A workaround is wrap this small Widget into Container with padding, and set background color to transparent and set the behavior of GestureDetector to HitTestBehavior.translucent.'),
          Text(
              'For example, the following wraps a 22*22 Widget into a bigger Invisible Container 44*44. But this way would change the Layout/Size.'),
          Text(
              'In the following example, we replace transparent with a light color for better demonstration.'),
          Padding(
            padding: const EdgeInsets.only(top: 10),
            child: Container(
              decoration: BoxDecoration(
                borderRadius: BorderRadius.all(
                  Radius.circular(5),
                ),
                border: Border.all(
                  color: Colors.black,
                  width: 1,
                ),
              ),
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: [
                  _buildLarge(),
                  _buildMedium(),
                  GestureDetector(
                    behavior: HitTestBehavior.translucent,
                    onTap: () => _showToast('22 is Tapped'),
                    child: Container(
                      color: Colors.redAccent.withOpacity(0.5),
                      padding: EdgeInsets.all(11),
                      child: _buildSmall(),
                    ),
                  ),
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildCell3() {
    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 10),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(
              'Another workaround is put a large transparent Container on the small Widget using Stack.'),
          Padding(
            padding: const EdgeInsets.only(top: 10),
            child: Container(
              decoration: BoxDecoration(
                borderRadius: BorderRadius.all(
                  Radius.circular(5),
                ),
                border: Border.all(
                  color: Colors.black,
                  width: 1,
                ),
              ),
              child: Stack(
                children: [
                  Row(
                    mainAxisAlignment: MainAxisAlignment.spaceBetween,
                    children: [
                      _buildLarge(),
                      _buildMedium(),
                      _buildSmall(),
                    ],
                  ),
                  Positioned(
                    right: 0,
                    top: 0,
                    width: 44,
                    height: 44,
                    child: GestureDetector(
                      behavior: HitTestBehavior.translucent,
                      onTap: () => _showToast('22 is Tapped'),
                      child: Container(
                        color: Colors.redAccent.withOpacity(0.5),
                      ),
                    ),
                  )
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildCell4() {
    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 10),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(
              'Using ExpandTapWidget can expand the tap area without change its size or layout.'),
          Padding(
            padding: const EdgeInsets.only(top: 10),
            child: Container(
              decoration: BoxDecoration(
                borderRadius: BorderRadius.all(
                  Radius.circular(5),
                ),
                border: Border.all(
                  color: Colors.black,
                  width: 1,
                ),
              ),
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: [
                  _buildLarge(),
                  _buildMedium(),
                  ExpandTapWidget(
                    onTap: () => _showToast('22 is Tapped'),
                    tapPadding: EdgeInsets.fromLTRB(78, 11, 0, 11),
                    child: _buildSmall(),
                  ),
                ],
              ),
            ),
          ),
          Padding(
            padding: const EdgeInsets.only(top: 10),
            child: Text(
              'The tap area of above small red widget is illustrated as Cyan colored area.',
              style: TextStyle(
                fontStyle: FontStyle.italic,
                color: Colors.cyan,
              ),
            ),
          ),
        ],
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: LayoutBuilder(
        builder: (_, constraints) {
          return SingleChildScrollView(
            physics: AlwaysScrollableScrollPhysics(),
            child: ConstrainedBox(
              constraints: BoxConstraints(minHeight: constraints.maxHeight),
              child: Container(
                padding: EdgeInsets.symmetric(horizontal: 20),
                child: Column(
                  children: <Widget>[
                    _buildCell1(),
                    Divider(
                      color: Colors.grey,
                    ),
                    _buildCell2(),
                    Divider(
                      color: Colors.grey,
                    ),
                    _buildCell3(),
                    Divider(
                      color: Colors.grey,
                    ),
                    _buildCell4(),
                    Padding(
                      padding: const EdgeInsets.symmetric(vertical: 20.0),
                      child: FlatButton(
                        onPressed: () {
                          Navigator.of(context).push(MaterialPageRoute(
                              builder: (context) => DemoPage()));
                        },
                        child: Text(
                          'Demo',
                          style: TextStyle(
                            color: Colors.white,
                          ),
                        ),
                        color: Colors.blue,
                      ),
                    ),
                  ],
                ),
              ),
            ),
          );
        },
      ),
    );
  }
}

class DemoPage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Demo'),
      ),
      body: LayoutBuilder(
        builder: (_, constraints) {
          return SingleChildScrollView(
            physics: AlwaysScrollableScrollPhysics(),
            child: ConstrainedBox(
              constraints: BoxConstraints(minHeight: constraints.maxHeight),
              child: Container(
                padding: EdgeInsets.all(20),
                child: Column(
                  children: <Widget>[
                    Container(
                      width: double.infinity,
                      margin: EdgeInsets.symmetric(vertical: 10),
                      child: Text(
                        "The Cyan colored area is used to illustrate the expand area. The size of all the following bigger Container is 100 * 100.",
                        style: TextStyle(fontStyle: FontStyle.italic),
                      ),
                    ),
                    Container(
                      width: double.infinity,
                      margin: EdgeInsets.symmetric(vertical: 10),
                      child: Text(
                          'The size of blue Container is 50 * 50, tapPadding is EdgeInsets.all(25).'),
                    ),
                    Container(
                      alignment: Alignment.center,
                      width: 100,
                      height: 100,
                      child: ExpandTapWidget(
                        onTap: () => _showToast('Tap Blue'),
                        tapPadding: EdgeInsets.all(25),
                        child: Container(
                          width: 50,
                          height: 50,
                          color: Colors.blue,
                        ),
                      ),
                    ),
                    Container(
                      width: double.infinity,
                      margin: EdgeInsets.symmetric(vertical: 10),
                      child: Text(
                        "Notice: \nThe expand tap area won't exceed the size of parent box. In this case, tapPadding is EdgeInsets.all(100).",
                        style: TextStyle(fontWeight: FontWeight.w600),
                      ),
                    ),
                    Container(
                      alignment: Alignment.center,
                      width: 100,
                      height: 100,
                      child: ExpandTapWidget(
                        onTap: () => _showToast('Tap Blue'),
                        tapPadding: EdgeInsets.all(100),
                        child: Container(
                          width: 50,
                          height: 50,
                          color: Colors.blue,
                        ),
                      ),
                    ),
                    Container(
                      width: double.infinity,
                      margin: EdgeInsets.symmetric(vertical: 10),
                      child: Text(
                          "Parent: GestureDetector, Child: ExpandTapWidget"),
                    ),
                    GestureDetector(
                      onTap: () => _showToast('Tap Yellow'),
                      child: Container(
                        width: 100,
                        height: 100,
                        alignment: Alignment.center,
                        color: Colors.yellow,
                        child: Container(
                          alignment: Alignment.center,
                          width: 70,
                          height: 70,
                          child: ExpandTapWidget(
                            onTap: () => _showToast('Tap Blue'),
                            tapPadding: EdgeInsets.all(10),
                            child: Container(
                              width: 50,
                              height: 50,
                              color: Colors.blue,
                            ),
                          ),
                        ),
                      ),
                    ),
                    Container(
                      width: double.infinity,
                      margin: EdgeInsets.symmetric(vertical: 10),
                      child: Text(
                          "Parent: ExpandTapWidget, Child: GestureDetector"),
                    ),
                    Container(
                      alignment: Alignment.center,
                      width: 100,
                      height: 100,
                      child: ExpandTapWidget(
                        onTap: () => _showToast('Tap Blue'),
                        tapPadding: EdgeInsets.all(100),
                        child: Container(
                          width: 50,
                          height: 50,
                          color: Colors.blue,
                          alignment: Alignment.center,
                          child: GestureDetector(
                            onTap: () => _showToast('Tap Yellow'),
                            child: Container(
                              color: Colors.yellow,
                              width: 30,
                              height: 30,
                            ),
                          ),
                        ),
                      ),
                    ),
                  ],
                ),
              ),
            ),
          );
        },
      ),
    );
  }
}
